### ISSUE TITLE
revisit: browser.urlbar.maxRichResults

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#125]
@nodiscc says (https://github.com/pyllyukko/user.js/commit/116d9863de4b4b946d699368bebd7e7dff0b7903#commitcomment-22223673)

> * =12: when typing a URL in the address bar, a maximum of 12 bookmarks or history items are suggested in the dropdown below the URL bar
> * =0: nothing is shown in the dropdown

So I propose setting up a definitive test with steps to be reproduced, with screenshots of results, to be used in vanilla Firefox. I can test this on Windows. I would think that crssi, earthlng and my own results from earlier testing in which we all found `browser.urlbar.autocomplete.enabled` = 0 did not hide any history-bookmark-opentab suggestions was good.

But to be sure, and to check the behavior is not different in linux, let's document it properly.

**Comment by earthlng** [comment]
The commit is about `browser.urlbar.maxRichResults`, isn't it?
`browser.urlbar.autocomplete.enabled` does autocomplete on the urlbar when you type something.

**Comment by crssi** [comment]
`browser.urlbar.autocomplete.enabled` name is somehow wrong, since when this is false, the search actually doesn't work.

**Comment by earthlng** [comment]
> `browser.urlbar.autocomplete.enabled` name is somehow wrong, since when this is false, the search actually doesn't work.

Can confirm. Weird! Thanks for pointing that out mate!

> and I have not seen that when true any auto complete would do some magic

If I type in a "g" for example then it autocompletes to ghacks.net for me, because I have ghacks.net in either my bookmarks or history and it has the most "points" so to speak. The points are accumulated based on the "bonuses" from `places.frecency.*Bonus`. If I wanted "g" to autocomplete to github.com for example I could reload github.com in a normal window a couple times until its bonuses are higher than ghacks.net.

**Comment by Thorin-Oakenpants** [comment]
sorry - see rename of topic - my bad, copypasta malfunction

nodiscc said that `browser.urlbar.maxRichResults` controls his dropdown for history+bookmark entries shown and if he sets it to zero then it disables the dropdown. I said we tested this and all three (or 4) of us all had the same results that (at least in Windows) it did no such thing (I think we worked out anything under 10 didn't work).

I will post my step by step test methods and results, and pictures and ask nodiscc to replicate.

**Comment by crssi** [comment]
and more than 10 also didn't make sense

**Comment by Thorin-Oakenpants** [comment]
Steps to reproduce
---test1
1. Nilla FF/profile [Tested with 53 and 51, since it was around 51 when we did our earlier tests]
2. defaults
   * browser.urlbar.suggest.history = true
   * browser.urlbar.suggest.bookmark = true
   * browser.urlbar.suggest.openpage = true
   * browser.urlbar.maxRichResults = 10
3. go to github, bookmark 5 pages, visit another 5 (I just used the items in the header _ footer nav sections)
4. close all tabs but keep FF open
5. type `git` in the urlbar
6. results
---test2
7. change browser.urlbar.maxRichResults = 0
8. type `git` in the urlbar
9. results

![maxrich](https://cloud.githubusercontent.com/assets/16656956/26313329/a3146fd4-3f5e-11e7-993f-576a792e5f81.png)

Results as follows:
- Test1: value 10: 1st line is taken by search with [search engine loaded in searchbar], the other 9 items are filled by ALL my 5 bookmarks and 4 history items
- Test2: value 0: 1st line taken by search with [search engine loaded in searchbar], all the rest is not displayed

Exact same result in FF51 and FF53 just in case something changed.

Conclusion: maxRichResults does control the number of entries from history/bookmarks/openpages and setting it to zero does disable it.

So how did ALL three (four?) of us come to the same conclusion that maxRichResults does not work? See https://github.com/ghacksuserjs/ghacks-user.js/issues/63#issuecomment-287537248 etc and 0800 redo issue etc.

**Comment by earthlng** [comment]
We tested it in FF52 at the time I think, maybe they fixed it again in FF53 idk

**Comment by crssi** [comment]
I have the same setting as I had here [#63 (comment)](https://github.com/ghacksuserjs/ghacks-user.js/issues/63#issuecomment-287537248)
and the results are the same... 0 does not disable it at all, tested just now.

**Comment by Thorin-Oakenpants** [comment]
Quote crssi : https://github.com/ghacksuserjs/ghacks-user.js/issues/63#issuecomment-287537248
> Open a few tabs with for example ghacks github, have a few bookmarks with github.
Set maxRichResults = 0, type into URLbar "github", you will see that 0 does not disables it.

earthlng
> I played with it and see what you mean. It behaves more like a MIN than a MAX value (now?).
Even when I set it to 1 it sometimes shows 3 or 4 or whatever, depending on the amount of bookmarks and/or history

me
> So I fired up a clean FF52, defaults for all five 0850s, added some quick history and bookmarks .. blah blah blah .. same as you guys. This does nothing to stop suggestions. I only tried with the three in 0850c, didn't bother to try with combos involving search in urlbar.

That's three of us who could not stop the display (but earthlng didn't say he tried 0). So what gives? Does some other setting affect the behavior?

**Comment by Thorin-Oakenpants** [comment]
> We tested it in FF52 at the time I think, maybe they fixed it again in FF53 idk

OK, I'll test it in Nilla 52

**Comment by earthlng** [comment]
0 also doesn't work for me in FF53

**Comment by Thorin-Oakenpants** [comment]
52.01 - same result. maxRichResults disables all results from history/bookmarks/openpages

**Comment by Thorin-Oakenpants** [comment]
> 0 also doesn't work for me in FF53

Do you mean in FF53 you set maxrich to 0 and this did NOT disable bookmark/history/openpage suggestions?

**Comment by earthlng** [comment]
yep, just now and crssi also said the same thing. no idea why though

**Comment by nodiscc** [comment]
I get the same results as @Thorin-Oakenpants in his [tests above](https://github.com/ghacksuserjs/ghacks-user.js/issues/125#issuecomment-303116422) in Firefox 45 ESR (Debian).

**Comment by Thorin-Oakenpants** [comment]
Man this is confusing. @crssi and earthlng - are you using a NILLA profile and ONLY changing maxrich for the two results?

**Comment by earthlng** [comment]
No, I'm using my main FF but I have the 3 `browser.urlbar.suggest.*` prefs you listed above set to true by default and maxrich doesn't do shit here

**Comment by Thorin-Oakenpants** [comment]
can you please test in a nilla earthlng - because I'm now starting to think other prefs influence maxRich

**Comment by Thorin-Oakenpants** [comment]
> No, I'm using my main FF but I have the 3 browser.urlbar.suggest.* prefs you listed above set to true by default and maxrich doesn't do shit here

I just tested in my main FF (which uses the ghacks.js exactly) - got the exact same results as theearlier tests - maxRich=0 does disable the suggestions

**Comment by earthlng** [comment]
> because I'm now starting to think other prefs influence maxRich

lol you think? :) that's what I have been saying above. It's clear that that's the case here

**Comment by earthlng** [comment]
your main FF has the 3 browser.urlbar.suggest.* prefs disabled though, right?

**Comment by Thorin-Oakenpants** [comment]
> your main FF has the 3 browser.urlbar.suggest.* prefs disabled though, right?

I did the two tests - that is I made sure in my main FF that the four prefs were at default, suggestions showed. flipped as per test 2 maxrich to 0 and suggestions did not show.

**Comment by Thorin-Oakenpants** [comment]
I going to go back to my test 53 from above and trying to see if any of the other 0800 urlbar/searchbar prefs change the behaviour

**Comment by earthlng** [comment]
> you're not messing with the dropdown in any way in the css right?

correct, no custom css for that part

**Comment by Thorin-Oakenpants** [comment]
Sidenote: I'm really only interested in whether 0 disables the suggestions, but just to let you know, in my test profile I have at least 10 github entries, 5 bookmarks and a swag of history. I can get the values 1, or 2, or 3, etc right up to 10 to display that EXACT number of items in the dropdown

OK, so I went back to my 53 test profile for this and did the repeated the same tests, but added a third pref for each one - namely all the other 0800 prefs such as browser.search.suggest.enabled, browser.urlbar.suggest.searches, browser.urlbar.autocomplete.enabled, browser.urlbar.autoFill, browser.urlbar.autoFill.typed, browser.urlbar.oneOffSearches <- that is I tried with each one set to true and false, reset it, moved on to the next one.

Every single test showed maxRich works as expected - that is it does limit the number in the dropdown, and thus 0 disables it. So at this stage I can't see another pref interfering with it.

I'm stumped

**Comment by earthlng** [comment]
> that is I tried with each one set to true and false, reset it, moved on to the next one.

thx for the testing. It must be a combination of prefs then because it clearly doesn't work for me or crssi

**Comment by Thorin-Oakenpants** [comment]
Yup .. thats the only logical next conclusion .. a combo of prefs. I was going to look at the dxr code but I'm a bit sh*t at reading that stuff .. i miss COBOL

**Comment by earthlng** [comment]
Ok so with a copy of my FF53 profile in a portable FF53 and with all my addons removed, I still get the same results i.e. maxrich is ignored. browser.urlbar.autocomplete.enabled;false however stops the dropdown completely. I'll try to figure out how to get an EXACT match with maxrich.

**Comment by earthlng** [comment]
With maxrich=0 it works for some characters but not for others. for example "a", "b", "c" all only show the visit with/search with, but "g" and "h" shows some entries. And it's the same in my main FF.
Something in the places file seems to influence how many entries are shown. Maybe something to do with the bonuses I mentioned above.

**Comment by crssi** [comment]
same as @earthlng and I also didn't do any css customs

**Comment by earthlng** [comment]
So if you use PB a lot, chances are that none of your bookmarks or history will have high bonuses and therefore maxrich=0 seems to work correctly but if you do have entries with high bonuses then maxrich=0 is ignored for those entries

**Comment by Thorin-Oakenpants** [comment]
> browser.urlbar.autocomplete.enabled;false however stops the dropdown completely.

Listen up. Open two about:config tabs. In one enter `browser.urlbar.autocomplete.enabled` and in the other enter `browser.urlbar.suggest.` (for the three prefs)

Set all four prefs as true. Now change `browser.urlbar.autocomplete.enabled` to true and your three *suggest prefs will auto change to true. Now change `browser.urlbar.autocomplete.enabled` to false, and your three *suggest prefs will auto-change to false

Now the reverse. Change at least one *suggest pref and `browser.urlbar.autocomplete.enabled` will become true. Set all three *suggest prefs to false and `browser.urlbar.autocomplete.enabled` will become false

**Comment by earthlng** [comment]
Oh okay, yeah that explains why nothing is displayed with browser.urlbar.autocomplete.enabled;false

**Comment by Thorin-Oakenpants** [comment]
Heh .. no idea what happens if you start your FF with user..js values that clash eg `browser.urlbar.autocomplete.enabled`-true but all three *suggest as false. :) It probably corrects itself

**Comment by earthlng** [comment]
But I think we should comment out browser.urlbar.autocomplete.enabled because as @crssi mentioned the search "keywords" don't work otherwise. That's not the same as keyword.enabled;false btw. - the search keywords still work regardless of keyword.enabled

**Comment by Thorin-Oakenpants** [comment]
oh no .. don't confuse it more - start a new issue. WTF is "keywords"

**Comment by earthlng** [comment]
> WTF is "keywords"

Options > Search > there you can define "keywords" for use in the urlbar for every search engine

**Comment by crssi** [comment]
I don't use keywords.
But did a test your observation now... funny if you set in user.js
`user_pref("browser.urlbar.suggest.bookmark", false);`
`user_pref("browser.urlbar.suggest.openpage", false);`

and after those you do:
`user_pref("browser.urlbar.autocomplete.enabled", true);`

The both `browser.urlbar.suggest.` are preserved to false even that autocomplete is set to true.

**Comment by Thorin-Oakenpants** [comment]
> Heh .. no idea what happens if you start your FF with user..js values that clash eg `browser.urlbar.autocomplete.enabled`-true but all three *suggest as false. :) It probably corrects itself

OK, so I deliberately set a clash in my js, quite FF, started FF. I set the same as in the above quote. I never even touched the urlbar, didn't give it focus or anything. I opened about:config from a bookmark and checked the values, and because `browser.urlbar.autocomplete.enabled` was true, all three of my *suggest prefs were also true, despite the js setting them as false.

So I think we can need to do something with 0850a (urlbar*) and 0850c (*suggest) - merge them.

> How do you search Pants?

I have no keywords in any of my custom search engines. I exclusively use the searchbar. Maybe THIS is the difference?

**Comment by Thorin-Oakenpants** [comment]
The correlation is:
- If at least ONE *suggest is set to  true then *urlbar becomes true
- If *urlbar is set to false then all three *suggests become false
- You can still have *urlbar = true and a combination of *suggest values

**Comment by earthlng** [comment]
> Maybe THIS is the difference?

Maybe, but I don't think so. Have you tried in your Nilla FF's to do the bonuses thingy I wrote earlier?
I'm pretty sure that's what's causing it

**Comment by earthlng** [comment]
> I have no keywords in any of my custom search engines

you should try it. Makes it a lot easier to search directly where you want to search. Fe. I have "y" for youtube, "s" for startpage, "x" for ixquick, "w" for wikipedia etc

**Comment by Thorin-Oakenpants** [comment]
You'll have to explain how to do the bonuses thingy. I tried using different letters, so I racked a swag of amazon urls and tested `amaz` .. everything worked perfectly for maxRich still. I cannot for the life of me break it like I did back in 52

First I want to clean up 0850a + c. Something like this (note I swapped 0850b + c from the current)

```js
/* 0850a: disable location bar autocomplete
 * [NOTE] If this is false, all of 0850b will be FORCED to false as well ***/ <--added
user_pref("browser.urlbar.autocomplete.enabled", false);
/* 0850b: disable location bar suggestion types
 * [SETTING] Options>Privacy>Location Bar>When using the location bar, suggest
 * [NOTE] If any of these are true, 0850a will be FORCED to true ***/  <-- reworded
user_pref("browser.urlbar.suggest.history", false);
user_pref("browser.urlbar.suggest.bookmark", false);
user_pref("browser.urlbar.suggest.openpage", false);

/* 0850c: disable location bar dropdown
 * This value used to control the maximum number of entries that could appear in the location
 * bar dropdown. This is no longer the case, and thus, it does *NOT* hide any suggestions ***/
   // user_pref("browser.urlbar.maxRichResults", 0);
```

**Comment by earthlng** [comment]
> So I think we can need to do something with 0850a (urlbar*) and 0850c (*suggest) - merge them.

or remove 0850a because it seems to be the master switch and will be toggled automatically depending on 0850c

**Comment by Thorin-Oakenpants** [comment]
^^ see above edited draft thingy

**Comment by earthlng** [comment]
> You'll have to explain how to do the bonuses thingy.

I already did but okay let me try again. Assuming you still have the github pages in your test profile history/bookmarks that you pictured above. The last one in that list is "Learn how developers build software" so now you load that page over and over again until it makes it to the top of the dropdown list. Now that history entry will have a pretty high "frecency". Maybe do this for 2 or 3 pages that were previously at the bottom of the list. Now set maxrich=0 and type "g" or "git" again - if my theory is correct those pages should now get listed and effectively ignore maxrich=0.

**Comment by Thorin-Oakenpants** [comment]
ok - will try that. Yup, my history and bookmarks were all loaded just the once. They all have a visit count of 1 at the moment.

**Comment by earthlng** [comment]
~~keyword blah blah is disabled~~ >> search engine keywords are disabled
Yes, I think that's a good change ie your draft

**Comment by earthlng** [comment]
> They all have a visit count of 1 at the moment.

It's not the visit count that matters. One of my bookmarks that always shows up in the dropdown even with maxrich=0 has a visit count of 0!

**Comment by Thorin-Oakenpants** [comment]
Yeah, I was just saying the frequency of them was one visit and they were all equal.

OK, I took two of the maxrich results from the image in the test. I used the bottom two, doesn't matter. I visited those two pages around 10 times each. I did this from the suggestions, not from my bookmarks (they were history items anyway). Each page load I would close the tab and type `git` in a new tab and repeat selecting a suggestion. Voila .. check the picture. PROOF we're not all mad

![fuckbar](https://cloud.githubusercontent.com/assets/16656956/26320554/46354fe0-3f77-11e7-8e8a-d2ab5a396f07.png)

**Comment by Thorin-Oakenpants** [comment]
OK, explain to me what triggers a bonus. I never use suggestions because they have always been hidden. I have almost no bookmarks (about 50 I rarely use), because my 400 odd main ones are in a speed dial, but I visit stuff ALL the time. BUT I wipe history when I close. But I assume that if I leave my browser open long enough and load the same page numerous times over that period, it would gain enough bonuses, which would explain the test results I got in 52.

So is this a bug? But explain to me the bonus crap and where it's kept

**Comment by earthlng** [comment]
Afaik, PB doesn't update the bonuses unless you bookmark a page I think. Look at `places.*Bonus` in about:config. Idk when each one of those triggers but if you don't want the bonuses you can set all of those prefs to 0 and reset the "frecency" columns in places.sqlite.
I guess when you wipe the history, a lot of it will be cleared as well. Idk how it all works exactly just that it exists and that it determines what gets listed and what doesn't

**Comment by Thorin-Oakenpants** [comment]
Sounds like a ~~bug~~ feature to me!! I don't care about it, just as long as I document it.

I also need to know how the F do you add a keyword to a search engine (mine are all custom from Add To Search Bar, which I no longer have).

**Comment by earthlng** [comment]
> how the F do you add a keyword to a search engine

right there in Options > Search -> double click the "keyword" column and type in a keyword. It can be a single letter, a number or a string.

**Comment by Thorin-Oakenpants** [comment]
Ohh fark ... lulz .. must be FT DEEP dark or something .. nothing shows that was editable at all, not a lick, no column lines, nada ... tried everything EXCEPT a double click

**Comment by Thorin-Oakenpants** [comment]
This needs a clean up - wanna give it a go?

```js
/* 0850a: disable location bar autocomplete [controlled by 0850b]
 * [NOTE] If this is false, all of 0850b will be FORCED to false as well
 * [WARNING] If this is false, search engine keywords are disabled ***/
   // user_pref("browser.urlbar.autocomplete.enabled", false);

/* 0850b: disable location bar suggestion types [controls 0850a]
 * [SETTING] Options>Privacy>Location Bar>When using the location bar, suggest
 * [NOTE] If any of these are true, 0850a will be FORCED to true
 * and if all three are false, 0850a will be FORCED to false
 * [WARNING] If 0850a is false, search engine keywords are disabled ***/
user_pref("browser.urlbar.suggest.history", false);
user_pref("browser.urlbar.suggest.bookmark", false);
user_pref("browser.urlbar.suggest.openpage", false);

/* 0850c: disable location bar dropdown
 * This value control the total number of entries to appear in the location bar dropdown.
 * Accepted values are 0 to 10. A value of zero effectively disables the dropdown.
 * HOWEVER, items (bookmarks/history etc) with a high "frequency"/"bonus" will always
 * be displayed (no we do not know how these are calculated or what the threshold is).
 * [NOTE] Use this if you 0850a ends up being true to enable search engine keywords,
 * but you don't want suggestions  ***/
   // user_pref("browser.urlbar.maxRichResults", 0);
```

**Comment by earthlng** [comment]
Looks pretty good if you fix the typos ;)

* This value control**s**
* Use this if you 0850a ends up being true
* ~~A value of zero effectively disables the dropdown~~ 0 disables the dropdown
   * short and sweet but yours is fine too
* Accepted values are 0 to 10 **(default)**
  * ^^ maybe? or zero fucks given? :)

**Comment by Thorin-Oakenpants** [comment]
I tested it in my nilla, nothing over 10 works

**Comment by earthlng** [comment]
It works if you use custom css to remove the space limitation. But I don't think we need to go there

**Comment by Thorin-Oakenpants** [comment]
yeah, I know that, you've mentioned it a few times. No need to go to that detail. I just don;t want people saying, but I put in 15 and it doesnt work

**Comment by Atavic** [comment]
```
 * HOWEVER, items (bookmarks/history etc) with a high "frequency"/"bonus" will always
 * be displayed (no we do not know how these are calculated or what the threshold is).
```
That's the "Frecency" score that defines urls that should appear in the drop-down:

`places.frecency.*`

**Comment by Thorin-Oakenpants** [comment]
Hah .. I though that was a typo

**Comment by Atavic** [comment]
No, it's a fast telemetry thing. I do not like it as its monitoring times are set various ranges from minutes up to months.

**Comment by earthlng** [comment]
> Hah .. I though that was a typo

Lol, you must think I'm retarded. I mean once is typo but 4 times?? bitch please :)

**Comment by Thorin-Oakenpants** [comment]
```js
/* 0850a: disable location bar autocomplete [controlled by 0850b]
   // user_pref("browser.urlbar.autocomplete.enabled", false);

/* 0850b: disable location bar suggestion types [controls 0850a]
 * [SETTING] Options>Privacy>Location Bar>When using the location bar, suggest
 * [NOTE] If any of these are true, 0850a will be FORCED to true
 * and if all three are false, 0850a will be FORCED to false
 * [WARNING] If all three are false, search engine keywords are disabled ***/
user_pref("browser.urlbar.suggest.history", false);
user_pref("browser.urlbar.suggest.bookmark", false);
user_pref("browser.urlbar.suggest.openpage", false);

/* 0850c: disable location bar dropdown
 * This value controls the total number of entries to appear in the location bar dropdown.
 * [NOTE] Items (bookmarks/history/openpages) with a high "frecency"/"bonus" will always
 * be displayed (no we do not know how these are calculated or what the threshold is),
 * and this does not affect the search by search engine suggestion (see 0808)
/* [USAGE] This setting is only useful if you want to enable search engine keywords
   (i.e at least one of 0850b must be true) but you don't want suggestions to show ***/
   // user_pref("browser.urlbar.maxRichResults", 0);
```

thumbs up if I can commit (minus the extra line gaps)

**Comment by earthlng** [comment]
I think a better way to achieve what you wrote under 0850c [USAGE] is to only set .openpage to true.

**Comment by Thorin-Oakenpants** [comment]
^^ too much work. Some people have hundreds of open tabs, others dont keep history, others like me have almost no bookmarks (I use a speed dial) - so no one answer fits all as to which to enable to produce the least

**Comment by Atavic** [comment]
Info on GlobalFrecency [entries](https://wiki.mozilla.org/User:Sspitzer/GlobalFrecency) & Wilders [post](https://www.wilderssecurity.com/threads/firefox-31-released.366329/page-2#post-2395461) tell how to disable:

```
 browser.cache.frecency_experiment -1
```

**Comment by earthlng** [comment]
@Atavic that's for the cache not the bookmarks/history. To disable those you can set all the `places.frecency.*Bonus` prefs to 0.

**Comment by Thorin-Oakenpants** [comment]
Thanks @nodiscc for bringing this up :+1: :heart:

**Comment by Thorin-Oakenpants** [comment]
Ch.. ch.. changes! https://bugzilla.mozilla.org/show_bug.cgi?id=1264988 (FF56)

Hah! The popup doesn’t show a scrollbar anymore, instead it just adapts itself to the browser.urlbar.maxRichResults pref value
